py-polars/polars/internals/dataframe/frame.py

```diff
@@ -1255,144 +1173,219 @@ def __getitem__(
-            return self[range_to_slice(item)]
-
-        # df[:]
-        if isinstance(item, slice):
-            return PolarsSlice(self).apply(item)
-
-        # select rows by numpy mask or index
-        # df[np.array([1, 2, 3])]
-        # df[np.array([True, False, True])]
-        if _NUMPY_AVAILABLE and isinstance(item, np.ndarray):
-            if item.ndim != 1:
-                raise ValueError("Only a 1D-Numpy array is supported as index.")
-            if item.dtype.kind in ("i", "u"):
-                # Numpy array with signed or unsigned integers.
-                return self._from_pydf(
-                    self._df.take_with_series(self._pos_idxs(item, dim=0)._s)
+        ``polars`` abides by the philosophy that columns represent the primary
+        logical unit of a ``DataFrame``. With this in mind, ``df[X]`` always
+        selects *columns*. For any ``X``, it is shorthand for ``df[:, X]``.
+        Rows should be accessed explicitly with``df[X, :]``.
+
+        Note that when selecting by columns, the type of ``X`` fully determines the
+        return type of ``__getitem__``. If ``X`` is a scalar value (integer or
+        string) a ``Series`` is returned. Otherwise, a ``DataFrame`` is returned.
+        When selecting by rows, a ``DataFrame`` is always returned.
+
+        Additionally, ``df[X, int | str]`` is simply shorthand for ``df[int | str][X]``.
+        In other words, we first extract the column as a ``Series``, and then index
+        the row positions.
+
+        Examples
+        --------
+        >>> df = pl.DataFrame({"a": range(3), "b": range(3)})
+        >>> df["a"]
+        shape: (3,)
+        Series: 'a' [i64]
+        [
+            0
+            1
+            2
+        ]
+
+        >>> df[0]
+        shape: (3,)
+        Series: 'a' [i64]
+        [
+            0
+            1
+            2
+        ]
+
+        >>> df[["a"]]
+        shape: (3, 1)
+        ┌─────┐
+        │ a   │
+        │ --- │
+        │ i64 │
+        ╞═════╡
+        │ 0   │
+        ├╌╌╌╌╌┤
+        │ 1   │
+        ├╌╌╌╌╌┤
+        │ 2   │
+        └─────┘
+
+        >>> df[[0]]
+        shape: (3, 1)
+        ┌─────┐
+        │ a   │
+        │ --- │
+        │ i64 │
+        ╞═════╡
+        │ 0   │
+        ├╌╌╌╌╌┤
+        │ 1   │
+        ├╌╌╌╌╌┤
+        │ 2   │
+        └─────┘
+
+        >>> df[0, :]
+        shape: (1, 2)
+        ┌─────┬─────┐
+        │ a   ┆ b   │
+        │ --- ┆ --- │
+        │ i64 ┆ i64 │
+        ╞═════╪═════╡
+        │ 0   ┆ 0   │
+        └─────┴─────┘
+
+        >>> df[1:, "a":"b"]
+        shape: (2, 2)
+        ┌─────┬─────┐
+        │ a   ┆ b   │
+        │ --- ┆ --- │
+        │ i64 ┆ i64 │
+        ╞═════╪═════╡
+        │ 1   ┆ 1   │
+        ├╌╌╌╌╌┼╌╌╌╌╌┤
+        │ 2   ┆ 2   │
+        └─────┴─────┘
+
+        """
+        # df[X] == df[:, X]
+        if not isinstance(item, tuple):
+            return self[:, item]  # type: ignore[index]
+
+        rows, cols = item
```

Comment:
In the existing code, there was a check on that the length of the tuple should be two, otherwise it would be interpreted like a list was passed (of columns names for instance). I personally prefer tuples over lists (immutability), but this change would mean one could no longer do:
 
```python
df[("col1", "col2", "col3")]
```
I do realize that this, as is currently the case, that input types => return types would be broken, but I would be in favor of allowing the above.

Reply:
Hm, I can see the value in allowing tuples to be passed instead of lists (I too like immutability). But checking the length isn't enough.

```python
df = pl.DataFrame({"a": range(3), "b": range(3)})
```

`df[("a", "b")]` is invalid (both on my branch and on `master`). Should my `if` statement instead be something like

```python
if not isinstance(item, tuple) or (isinstance(item, tuple) and is_str_sequence(item)):
    ...
```

Reply:
I think so. But I do appreciate that others may have different opinions as this is a bit dirty in terms of implementations. So I will let others comment. @stinodego  @ritchie46